src/attributes.jl

```diff
@@ -449,6 +483,74 @@ struct RawParameter <: AbstractOptimizerAttribute
     name::Any
 end
 
+### Callbacks
+
+"""
+    struct OptimizeInProgress{AttrType<:AnyAttribute} <: Exception
+        attr::AttrType
+    end
+
+Error thrown from optimizer when `MOI.get(optimizer, attr)` is called inside an
+[`AbstractCallback`](@ref) while it is only defined once [`optimize!`](@ref) has
+completed. This is only defined if `is_set_by_optimize(attr)` is `true`.
+"""
+struct OptimizeInProgress{AttrType<:AnyAttribute} <: Exception
+    attr::AttrType
+end
+
+function Base.showerror(io::IO, err::OptimizeInProgress)
+    print(io, typeof(err), ": Cannot get result as the `MOI.optimize!` has not",
+          " finished.")
+end
+
+"""
+    abstract type AbstractCallback <: AbstractOptimizerAttribute end
+
+Abstract type for optimizer attribute representing a callback functions. The
+value set to subtypes of `AbstractCallback` is a function that may be called
+during [`optimize!`](@ref). As [`optimize!`](@ref) is in progress, the result
+attributes (i.e, the attributes `attr` such that `is_set_by_optimize(attr)`)
+may not be accessible from the callback hence trying to get result attributes
+might throw a [`OptimizeInProgress`](@ref) error.
+
+The value of the attribute should be a function taking only one argument, that
+is commonly called `callback_data`, that can be used for instance in
+[`LazyCallback`](@ref), [`HeuristicSolution`](@ref).
+"""
+abstract type AbstractCallback <: AbstractOptimizerAttribute end
+
+"""
+    LazyCallback() <: AbstractCallback
+
+The *lazy* callback can be used to submit [`LazyConstraint`](@ref) which are
+added on-demand at integer nodes in the branch and bound tree. Note that there
+is no guarantee that the callback is called at *every* integer primal solution.
+
+The solution at the integer node is accessed through
+[`VariablePrimalAtIntegerNode`](@ref). Trying to access other result
+attributes will throw [`OptimizeInProgress`](@ref) as discussed in
+[`AbstractCallback`](@ref).
+"""
+struct LazyCallback <: AbstractCallback end
+
+"""
+    HeuristicCallback() <: AbstractCallback
+
+The *heuristic* callback can be used to submit [`HeuristicSolution`](@ref) at
+fractional (i.e., non-integer) nodes in the branch and bound tree. Note that
```

Comment:
Are HeuristicCallbacks ever called at integer nodes asking for a better solution?

Reply:
In general, no.
See for instance the [CPLEX docs](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.9.0/ilog.odms.cplex.help/refcallablelibrary/cpxapi/callbackpostheursoln.html)